marshal/consumer.go

```diff
@@ -235,12 +235,47 @@ func (c *Consumer) defaultTopicPartitions() int {
 	return 0
 }
 
+// claimTerminated is called by a claim when they've terminated. This is used so we can
+// ensure topic claim semantics are adhered to. In topic claim mode this will be called
+// by every claim during a release.
+func (c *Consumer) claimTerminated(cl *claim, released bool) {
+	// For now, we don't care except in the topic claim mode
+	if !c.options.ClaimEntireTopic {
+		return
+	}
+
+	latestClaims, claimsErr := c.GetCurrentTopicClaims()
+	if claimsErr != nil {
+		// This only impacts update topic claims later
+		log.Errorf("claimTerminated: failed getting current claims: %s", claimsErr)
+	}
+	delete(latestClaims, cl.topic)
+
+	// This is a topic claim, so we need to perform the same operation on the rest of
+	// the claims in this topic
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	for _, claim := range c.claims[cl.topic] {
+		if cl != claim {
+			if released {
+				go claim.Release()
+			} else {
+				go claim.Terminate()
```

Comment:
Won't these calls to Release/Terminate result in this function getting called again immediately? We can end up trying to release each partition several times because we can release the lock before these calls to Release/Terminate take effect.